### PR TITLE
Add @CheckReturnValue to appropriate public APIs.

### PR DESCRIPTION
**Pull request by NightlyNexus**
Moshi's a good small library to try this out on.
If this looks good, we could file issues on other libraries to get these in.  If this doesn't look good, no time was lost.

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java [hunk 1]
```diff
@@ -30,13 +30,13 @@
  * Converts Java values to JSON, and JSON values to Java.
  */
 public abstract class JsonAdapter<T> {
-  public abstract @Nullable T fromJson(JsonReader reader) throws IOException;
+  @CheckReturnValue public abstract @Nullable T fromJson(JsonReader reader) throws IOException;
```
Comment (NightlyNexus) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: There may be some use case for having a delegate adapter that throws away a value you don't care about, but that seems weird enough that a suppression in app code with an explanation is warranted.

**Comment by adriancole** [comment]
maybe for the lazy c&p what exactly checkReturnValue implies?

**Comment by NightlyNexus** [comment]
@adriancole
It will give a hint in the IDE if you forget to use assign the result of the method call. (It's usually applied to factory methods and methods with no side effects.)

There's also an Error Prone check: http://errorprone.info/bugpattern/CheckReturnValue

Here's a real example where `@CheckReturnValue` prevents mistakes: https://github.com/square/moshi/pull/353

**Comment by NightlyNexus** [comment]
>This annotation is used to denote a method whose return value should always be checked after invoking the method.

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java (hunk 1 above)
Comment (swankjesse) [reply moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: definitely. `reader.skipValue()` is better.

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java [hunk 2]
```diff
@@ -143,7 +143,7 @@ public final void toJson(BufferedSink sink, @Nullable T value) throws IOExceptio
   }
 
   /** Returns a JSON adapter equal to this, but is lenient when reading and writing. */
-  public final JsonAdapter<T> lenient() {
+  @CheckReturnValue public final JsonAdapter<T> lenient() {
```
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: these ones are particularly useful

**Comment by adriancole** [comment]
thx for the details. looks very useful (though seems like could apply to
hordes of methods)

**Pull request merged**